git/odb/object_writer.go

```diff
@@ -0,0 +1,123 @@
+package odb
+
+import (
+	"compress/zlib"
+	"crypto/sha1"
+	"fmt"
+	"hash"
+	"io"
+	"sync/atomic"
+)
+
+// ObjectWriter provides an implementation of io.Writer that compresses and
+// writes data given to it, and keeps track of the SHA1 hash of the data as it
+// is written.
+type ObjectWriter struct {
+	// w is the underling writer that this ObjectWriter is writing to.
+	w io.Writer
+	// sum is the in-progress hash calculation.
+	sum hash.Hash
+
+	// wroteHeader is a uint32 managed by the sync/atomic package. It is 1
+	// if the header was written, and 0 otherwise.
+	wroteHeader uint32
+
+	// closeFn supplies an optional function that, when called, frees an
+	// resources (open files, memory, etc) held by this instance of the
+	// *ObjectWriter.
+	//
+	// closeFn returns any error encountered when closing/freeing resources
+	// held.
+	//
+	// It is allowed to be nil.
+	closeFn func() error
+}
+
+var _ io.WriteCloser = (*ObjectWriter)(nil)
+
+// nopCloser provides a no-op implementation of the io.WriteCloser interface by
+// taking an io.Writer and wrapping it with a Close() method that returns nil.
+type nopCloser struct {
+	// Writer is an embedded io.Writer that receives the Write() method
+	// call.
+	io.Writer
+}
+
+var _ io.WriteCloser = (*nopCloser)(nil)
+
+// Close implements the io.Closer interface by returning nil.
+func (n *nopCloser) Close() error {
+	return nil
+}
+
+// NewObjectWriter returns a new *ObjectWriter instance that drains incoming
+// writes into the io.Writer given, "w".
+func NewObjectWriter(w io.Writer) *ObjectWriter {
+	return NewObjectWriteCloser(&nopCloser{w})
+}
+
+// NewObjectWriter returns a new *ObjectWriter instance that drains incoming
+// writes into the io.Writer given, "w".
+//
+// Upon closing, it calls the given Close() function of the io.WriteCloser.
+func NewObjectWriteCloser(w io.WriteCloser) *ObjectWriter {
+	zw := zlib.NewWriter(w)
+	sum := sha1.New()
+
+	return &ObjectWriter{
+		w:   io.MultiWriter(zw, sum),
+		sum: sum,
+
+		closeFn: func() error {
+			if err := zw.Close(); err != nil {
+				return err
+			}
+			if err := w.Close(); err != nil {
+				return err
+			}
+			return nil
+		},
+	}
+}
+
+// WriteHeader writes object header information and returns the number of
+// uncompressed bytes written, or any error that was encountered along the way.
+//
+// WriteHeader MUST be called only once, or a panic() will occur.
+func (w *ObjectWriter) WriteHeader(typ ObjectType, len int64) (n int, err error) {
+	if !atomic.CompareAndSwapUint32(&w.wroteHeader, 0, 1) {
+		panic("git/odb: cannot write headers more than once")
```

Comment:
@technoweenie interesting question. I think I could go either way on this. My thought when implementing this was that if a program calls `Write()` before `WriteHeader()`, something is wrong at a higher level and any error returned from `Write()` or `WriteHeader()` would not be respected, so we should just terminate.

With respect to LFS specifically, all uses of the `*ObjectWriter` from the `*ObjectDatabase` type call these two in the correct order, so I think I'm inclined to leave this as-is.